### PR TITLE
[Fusion] Explain a fused hybrid score from its legs' explanations

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description

Adding `explain` for fused `hybrid` query. Structurally this is the `profile` mechanism of #1977 a second time, on the same approach and with the same hook: filter attaches a collector before the rewrite, fan-out fills it, filter's response listener reads from it. New DTO class FusedDocExplanations carry the report. 

Where we attach new explain classes, and where it deliberately do not

| Report | Attaches to | Why not more |
|---|---|---|
| `profile` (#1977) | every fused hybrid in the request | a profile section is a list, so N hybrids can each contribute |
| `explain` (this PR) | **only** the fused hybrid that *is* the request's own `query`, by reference identity | a hit's explanation is one tree describing that hit's score; nested, the fused score is an input to the enclosing query's scoring, so replacing the tree would describe the wrong thing |

It follows same main fusion/resolver paradigm - no pipeline or processor is required, specifically existing [explanation response processor](https://docs.opensearch.org/latest/search-plugins/search-pipelines/explanation-processor/). One more advantage over existing logic for explain - with this change we run actual query once and explain part only processes scores differently, so it's faster. Today explain in hybrid query runs query phase twice to get raw queries scores. 

---

#### Query examples

Index used by the ITs: 6 documents, ids `1`–`6`, each with the same `text` and a 2-dimensional `vec`, queried by a
`knn` leg and a `term` leg that matches all six.

#### 1. The flat case — every leg described under the fused score

```json
GET test-fused-explain/_search
{
  "explain": true,
  "track_total_hits": true,
  "size": 7,
  "query": {
    "hybrid": {
      "fusion": {
        "window_size": 10,
        "normalization": { "technique": "min_max" },
        "combination": { "technique": "arithmetic_mean" }
      },
      "queries": [
        { "knn":  { "vec": { "vector": [1.1, 1.0], "k": 10 } } },
        { "term": { "text": "hello" } }
      ]
    }
  }
}
```

Each ranked hit carries:

```json
"_explanation": {
  "value": 0.75,
  "description": "arithmetic_mean combination of:",
  "details": [
    {
      "value": 1.0,
      "description": "min_max normalization of:",
      "details": [
        { "value": 0.98, "description": "within top 10 docs", "details": [] }
      ]
    },
    {
      "value": 0.5,
      "description": "min_max normalization of:",
      "details": [
        { "value": 0.287, "description": "weight(text:hello in 0) [PerFieldSimilarity], result of:", "details": [ "..." ] }
      ]
    }
  ]
}
```

#### 2. Weighted combination — the weights are named, not folded in

```json
"combination": { "technique": "arithmetic_mean", "parameters": { "weights": [0.7, 0.3] } }
```

```json
"_explanation": {
  "value": 0.85,
  "description": "arithmetic_mean, weights [0.7, 0.3] combination of:",
  "details": [ "…leg 0, value 1.0…", "…leg 1, value 0.5…" ]
}
```

`0.7 × 1.0 + 0.3 × 0.5 = 0.85` — the weights apply to the values the children report, so the children stay readable
as the normalized leg scores they are.

#### 3. document matching one leg

```json
{ "knn": { "vec": { "vector": [1.1, 1.0], "k": 2 } } }
```

With `k: 2` the ANN leg contributes two candidates, so the other four documents reach fusion through the `term` leg
alone. Their trees carry **one** leg node, not two with a zero — the same choice classic makes.

#### 4. `rescore` — the fused score nests under the final score

```json
{
  "explain": true,
  "query": { "hybrid": { "fusion": { "...": "..." }, "queries": [ "..." ] } },
  "rescore": {
    "window_size": 10,
    "query": {
      "rescore_query": { "term": { "text": "hello" } },
      "query_weight": 1.0,
      "rescore_query_weight": 2.0
    }
  }
}
```

When the score round 2 returned is not the fused score, an extra node is inserted above the combination:

```json
"_explanation": {
  "value": 1.324,
  "description": "score of the fused hybrid query after post-fusion scoring, computed from:",
  "details": [
    { "value": 0.75, "description": "arithmetic_mean combination of:", "details": [ "..." ] }
  ]
}
```

`rescore` is the **only** way a fused hit's score can differ from its fused score: `boost` on a `hybrid` is rejected
at parse time (`[hybrid] query does not support [boost]`, HTTP 400).

### Related Issues
https://github.com/opensearch-project/neural-search/issues/1930

### Check List
- [X] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [X] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/neural-search/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by vibrantvarun — COMMENTED** [review]
Intense review — traced the explain path end-to-end (FusedDocExplanations, FusedExplanationMerger, fuseDetailed, recordExplanations, the identity attach, and the merge order). It's solid: fuse() now returns fuseDetailed().fused() so explained/unexplained scores can't diverge; legNormalizedScores is exactly what the combiner consumed; the top-level-only identity attach is correct; and profiles-then-explanations merge order is right. One LOW correctness inconsistency in the explain tree for an edge case (inline), plus one robustness nit on the cast. Approve-worthy after the first.

File: src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java [hunk 1]
```diff
@@ -333,14 +358,57 @@ private static RankedDocs computeRankedDocs(SearchHit[][] legHits, FusionSpec fu
...
+            List<FusedDocExplanations.LegContribution> contributions = new ArrayList<>(legNormalizedScores.size());
+            for (int leg = 0; leg < legNormalizedScores.size(); leg++) {
+                Float normalizedScore = legNormalizedScores.get(leg).get(key);
+                if (Objects.isNull(normalizedScore)) {
+                    // The leg did not match this document. Classic renders no node for it either.
+                    continue;
+                }
+                contributions.add(new FusedDocExplanations.LegContribution(leg, normalizedScore, legExplanations.get(leg).get(key)));
+            }
+            explanations.addDocument(key, ranked.scores()[doc], contributions);
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: This stores `ranked.scores()[doc]` as the combination-node value, but that value is post-`scoreAboveTail` (line 504), which floors a `0.0` fused score to a tiny positive epsilon (#1973). The node's leg children, however, are the normalized values, which arithmetic-mean to the **raw** `0.0`. Because the Top boost is also the floored score, `hitScore == fusedScoreByKey`, so `FusedDocExplanations.explain` returns the combination node directly (no `FINAL_SCORE` wrapper) — a tree whose parent value (~epsilon) its children don't produce (0.0).

Reachable exactly per #1973's own analysis: a `weights` entry of `0.0` for a doc matching only that leg, or `l2` over an all-zero leg (`MIN_SCORE == 0.0`).

Suggest passing the **raw** fused score instead:
```java
explanations.addDocument(key, fused.fused().get(key), contributions);
```
`fused` and `key` are already in scope and the value is non-null for any ranked doc. That makes the combination node consistent with its children **and** surfaces the floor naturally through the existing `FINAL_SCORE_DESCRIPTION` node (since `hitScore` (floored) would then differ from `fusedScore` (raw)) — which is what that "score after post-fusion scoring" node is for. Worth a test pinning a `weights:[0,1]` (or l2-zero-leg) doc's explanation.

File: src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java [hunk 2]
```diff
@@ -333,14 +358,57 @@ private static RankedDocs computeRankedDocs(SearchHit[][] legHits, FusionSpec fu
...
+        final List<Map<String, Explanation>> legExplanations,
+        final ScoreCombinationTechnique combination,
+        final ScalarNormalizer normalizer
+    ) {
+        if (legExplanations.stream().allMatch(Map::isEmpty)) {
+            return;
+        }
+        // Same format strings as classic: ScoreCombiner#explainByShard and ExplanationUtils#getDocIdAtQueryForNormalization.
+        explanations.combinationDescription(
+            String.format(Locale.ROOT, "%s combination of:", ((ExplainableTechnique) combination).describe())
```
Comment (vibrantvarun) [inline src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: nit/robustness: this cast to `ExplainableTechnique` is unguarded. All four current combination techniques implement it and classic hybrid does the identical cast, so it's safe today — but `ScoreCombinationTechnique` doesn't enforce `ExplainableTechnique`, so a future technique that omits it would throw a `ClassCastException` (500) only on explained requests. An `instanceof` guard (skip the fused breakdown, keep round-2's explanation) or folding `describe()` into the combination contract would make it future-proof.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: ack, Your second option is the right fix and it's free, since `describe()` is already a default method. I'll make
`ScoreCombinationTechnique extend ExplainableTechnique` and all three casts delete themselves

File: src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java (hunk 1 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: ack, and taking the fix

**Comment by martin-gaievski** [comment]
> Intense review — traced the explain path end-to-end (FusedDocExplanations, FusedExplanationMerger, fuseDetailed, recordExplanations, the identity attach, and the merge order). It's solid: fuse() now returns fuseDetailed().fused() so explained/unexplained scores can't diverge; legNormalizedScores is exactly what the combiner consumed; the top-level-only identity attach is correct; and profiles-then-explanations merge order is right. One LOW correctness inconsistency in the explain tree for an edge case (inline), plus one robustness nit on the cast. Approve-worthy after the first.

I take two action out of this comment:
1. javadoc in test is factually wrong. `HybridQuerySearchRequestFilterTests` claims test "pins that the one rebuild carries the very `SearchHit` instances over: a rebuild that copied them would drop the explanation written afterwards." **It pins no such thing** — under profiles-first the explanation is written *after*  the rebuild, so a copying rebuild would receive the write on the copies and the test would still pass. Reword to what it actually pins ("asking for both yields both") and point at `FusedLegProfileMergerTests.java:166` for the `assertSame` that does pin it.
2. **Coverage gap** `profile` + `explain` together is exercised only through mocked consumers in a unit
 test: `grep -c profile HybridQueryFusedModeExplainIT.java` = 0 and `grep -c explain HybridQueryFusedModeProfileIT` = 0. Adding `profile: true` to one existing explain IT case closes it.

File: src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java (hunk 1 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: Confirmed fixed in f365c1e132 — now records `fused.fused().get(key)`, so a floored-0.0 doc's combination node reads 0.0 (consistent with its leg children) under the FINAL_SCORE node showing the actual round-2 score. Tests cover it too. Thanks.

File: src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java (hunk 2 above)
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: Just flagging the interface change didn't land in f365c1e132 (that commit was Finding 1 only) — `ScoreCombinationTechnique` still doesn't extend `ExplainableTechnique` and the cast at line 396 is still there. Non-blocking nit, so no objection to merging without it, but noting it in case the follow-up got dropped.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/query/HybridFusionOrchestrator.java]: yes, my miss, pushed commit with the change

**Review by vibrantvarun — APPROVED** [review]
LGTM on 739d6c5cee (rebased clean off the earlier conflict). Both review findings resolved and re-verified at this head: the explain combination node now records the raw computed fused score (fused.fused().get(key)) so a floored-0.0 doc reads 0.0 under the FINAL_SCORE node — consistent with its leg children — and ScoreCombinationTechnique now extends ExplainableTechnique with every (ExplainableTechnique) cast removed, so an explained request can't 500 on a future technique. Explain path is sound overall: fuse() == fuseDetailed().fused() so scores can't diverge, legNormalizedScores is exactly what the combiner consumed, and explain attaches only to the top-level fused hybrid by identity. Thanks for the quick turnaround.
